File: supabase/functions/_shared/archetype.ts
// =====================================================================
// Project archetype classifier.
//
// Runs at intake to decide whether a submittal is inside the current
// pilot scope. Out-of-scope plans get rejected before the rule engine
// even runs — keeping the accuracy claim defensible.
//
// "Pilot scope" is a per-agency allowlist (agencies.pilot_archetypes
// JSONB array). If the column is empty or null, every archetype is
// accepted (legacy / unrestricted operation).
//
// Pure function — no LLM, no network. Inputs are the extracted scope,
// the resolved property overlay (if any), and the raw plan text for
// keyword cues. Reasoning is human-readable so reviewers can see why
// something was excluded.
// =====================================================================
import type { BuildingScope } from "./extract.ts";
import type { PropertyProfile } from "./property.ts";
import { PILOT_ARCHETYPES_DEFAULT } from "./pilot_config.ts";

export type ProjectArchetype =
  // In-pilot — Los Angeles
  | "la_sfr_typ_vb_ministerial"     // Single-family residence, Type V-B, no overlay flags
  | "la_ti_commercial"              // Commercial tenant improvement (no shell modification)
  // In-pilot — Ventura County (unincorporated)
  | "ventura_sfr_typ_vb_ministerial"  // SFR Type V-B outside VHFHSZ and outside Coastal Zone
  | "ventura_ti_commercial"           // Commercial TI in Ventura County, no shell change
  // Out-of-pilot (each maps to a specific human-readable reason)
  | "la_hillside_sfr"               // Subject to Hillside / BHO geometry rules
  | "la_hpoz_property"              // In a Historic Preservation Overlay Zone
  | "la_coastal_zone"               // CA Coastal Commission jurisdiction
  | "ventura_vhfhsz_sfr"            // Ventura County parcel in Very-High FHSZ (CBC Ch. 7A)
  | "ventura_ag_building"           // Agricultural-zoned structure, county-specific exemptions
  | "high_rise_or_mid_rise"         // Building height > 75 ft
  | "multifamily_new_construction"  // R-2/R-1 new build (not TI)
  | "mixed_use_new_construction"
  | "unclassified";                 // Couldn't tell — needs human triage

export interface ArchetypeResult {
  archetype: ProjectArchetype;
  in_pilot_scope: boolean;
  reasoning: string[];           // human-readable bullets explaining why
  excluded_overlays: string[];   // empty when in scope; lists which overlays kicked us out
}

// Default allowlist sourced from pilot_config.ts so a single edit there
// rewires the gate, the brief, and the eval harness simultaneously.
const IN_SCOPE: ReadonlySet<ProjectArchetype> = new Set(PILOT_ARCHETYPES_DEFAULT);

// =====================================================================
// classifyArchetype
//
// Order matters: we check "kicked out by overlay" cases first (cheapest
// reject), then identify the in-scope archetype, then fall back to
// unclassified.
// =====================================================================
export function classifyArchetype(
  scope: BuildingScope,
  property: PropertyProfile | null | undefined,
  planText: string,
  pilotArchetypes?: ProjectArchetype[],
): ArchetypeResult {
  const reasoning: string[] = [];
  const excluded: string[] = [];

  // ---- Jurisdiction detection (plan-text + parcel hint) --------------
  // Used by both LA and Ventura branches to pick the right out-of-scope
  // archetype slug when overlays fire.
  const venturaTextCue =
    /\bventura\s+county\b|\bcounty\s+of\s+ventura\b|\bcamarillo\b|\bsimi\s+valley\b|\bthousand\s+oaks\b|\boxnard\b|\bojai\b|\bmoorpark\b|\bport\s+hueneme\b|\bfillmore\b|\bsanta\s+paula\b/i;
  const isVenturaJurisdiction =
    /ventura/i.test(property?.parcel?.jurisdiction ?? "") ||
    venturaTextCue.test(planText);

  // ---- Hard rejects: address-derived overlays --------------------
  // These are the LA-specific accuracy killers that we deliberately
  // park out of pilot scope until we have ground-truth data on them.
  if (property?.parcel?.zoning_code) {
    const z = property.parcel.zoning_code.toUpperCase();
    if (/H$|HILLSIDE|\bBMO\b|\bBHO\b/i.test(z)) {
      excluded.push("Hillside / BHO zoning");
      reasoning.push(`Parcel zoning code "${z}" indicates Hillside overlay`);
      return finalize("la_hillside_sfr", excluded, reasoning, pilotArchetypes);
    }
  }
  if (property?.coastal_zone?.in_coastal_zone) {
    excluded.push("CA Coastal Zone");
    reasoning.push("Address is inside the CA Coastal Commission jurisdiction");
    return finalize("la_coastal_zone", excluded, reasoning, pilotArchetypes);
  }

  // ---- Ventura County VHFHSZ reject ------------------------------
  // Ventura's biggest accuracy risk: large unincorporated areas sit in
  // Very-High Fire Hazard Severity Zone, which pulls CBC Chapter 7A
  // wildfire-resistive-material requirements into the rule set. We
  // don't have enough ground truth on Ch. 7A material schedules yet,
  // so park these out of pilot until we do.
  if (isVenturaJurisdiction && property?.wui_zone?.in_wui) {
    excluded.push(`Ventura VHFHSZ (${property.wui_zone.haz_class})`);
    reasoning.push(
      `Address falls in CalFire ${property.wui_zone.haz_class} FHSZ — CBC Ch. 7A wildfire-resistive materials review out of pilot scope`,
    );
    return finalize("ventura_vhfhsz_sfr", excluded, reasoning, pilotArchetypes);
  }

  // Plan-text VHFHSZ cue (when GIS isn't resolved). Conservative —
  // only rejects on explicit Ventura + WUI/FHSZ co-mention.
  const wuiTextCue = /\bvery\s+high\s+fire\s+hazard\s+severity\s+zone\b|\bVHFHSZ\b|\bWUI\b|\bChapter\s+7A\b/i;
  if (isVenturaJurisdiction && wuiTextCue.test(planText)) {
    excluded.push("Ventura VHFHSZ (plan-text cue)");
    reasoning.push("Plan text references Ventura County WUI / Very-High FHSZ");
    return finalize("ventura_vhfhsz_sfr", excluded, reasoning, pilotArchetypes);
  }

  // ---- Ventura County agricultural building reject ---------------
  // Ag-zoned structures get county-specific exemptions (Title 8 ag
  // overlay + CBC §312 utility/misc). Outside the pilot until we
  // have fixtures with reviewer-confirmed exemption decisions.
  const agTextCue = /\bagricultural\s+(building|structure|exempt)|\bA-E\b|\bAE-\d|\bopen\s+space\s+agricultural\b/i;
  if (isVenturaJurisdiction && agTextCue.test(planText)) {
    excluded.push("Ventura agricultural building");
    reasoning.push("Plan text indicates an agricultural-zoned structure (county-specific exemptions apply)");
    return finalize("ventura_ag_building", excluded, reasoning, pilotArchetypes);
  }

  // ---- Plan-text overlay cues (when no property profile) ---------
  // The plan set itself often calls out applicable overlays even when
  // GIS isn't resolved. Conservative regex — we only reject on explicit
  // mention.
  const hillsideCues = /\bbaseline\s+hillside|\bBHO\b|\bhillside\s+ordinance\b|\bmulholland\b.*\bspecific\s+plan|RE40-?1H|RE15-?1H/i;
  if (hillsideCues.test(planText)) {
    excluded.push("Hillside / BHO (plan-text cue)");
    reasoning.push("Plan text references Baseline Hillside Ordinance or Hillside zoning");
    return finalize("la_hillside_sfr", excluded, reasoning, pilotArchetypes);
  }
  const hpozCues = /\bHPOZ\b|\bhistoric\s+preservation\s+overlay/i;
  if (hpozCues.test(planText)) {
    excluded.push("HPOZ");
    reasoning.push("Plan text references HPOZ (Historic Preservation Overlay Zone)");
    return finalize("la_hpoz_property", excluded, reasoning, pilotArchetypes);
  }

  // ---- Building scale rejects ----------------------------------------
  if (scope.height_ft != null && scope.height_ft > 75) {
    excluded.push("High-rise (> 75 ft)");
    reasoning.push(`Building height ${scope.height_ft} ft exceeds 75 ft high-rise threshold`);
    return finalize("high_rise_or_mid_rise", excluded, reasoning, pilotArchetypes);
  }
  if (scope.stories_above != null && scope.stories_above >= 5) {
    excluded.push("Mid-rise (≥ 5 stories)");
    reasoning.push(`${scope.stories_above} stories puts this above the pilot single-family / TI scope`);
    return finalize("high_rise_or_mid_rise", excluded, reasoning, pilotArchetypes);
  }

  // ---- In-scope identification ---------------------------------------
  // TI = explicit "tenant improvement" / interior alteration cues
  const tiCues = /\btenant\s+improvement\b|\bTI\b\s+plans?|interior\s+alteration|interior\s+remodel|\bsuite\s+\d/i;
  if (tiCues.test(planText) && !scope.mixed_occupancy) {
    if (isVenturaJurisdiction) {
      reasoning.push("Ventura County commercial tenant improvement");
      return finalize("ventura_ti_commercial", excluded, reasoning, pilotArchetypes);
    }
    reasoning.push("Plan text identifies this as a commercial tenant improvement");
    return finalize("la_ti_commercial", excluded, reasoning, pilotArchetypes);
  }

  // SFR Type V-B = R-3, single occupancy, V-B construction, modest scale
  const isR3 = scope.occupancies.includes("R-3") || scope.occupancy_primary === "R-3";
  const isVB = scope.construction_type === "V-B";
  const modestArea = (scope.per_story_area_sf ?? scope.building_area_sf ?? 0) <= 10_000;
  const lowRise = scope.stories_above == null || scope.stories_above <= 3;

  if (isR3 && isVB && modestArea && lowRise) {
    if (isVenturaJurisdiction) {
      reasoning.push("Ventura County single-family R-3 / Type V-B, ≤ 3 stories, no VHFHSZ or ag overlay");
      return finalize("ventura_sfr_typ_vb_ministerial", excluded, reasoning, pilotArchetypes);
    }
    reasoning.push("Single-family R-3 / Type V-B with modest area and ≤ 3 stories");
    return finalize("la_sfr_typ_vb_ministerial", excluded, reasoning, pilotArchetypes);
  }

  // Multifamily new construction = R-1 or R-2 + no TI cue
  if ((scope.occupancies.includes("R-1") || scope.occupancies.includes("R-2")) && !tiCues.test(planText)) {
    reasoning.push("R-1 / R-2 occupancy without TI markers — multifamily new construction");
    return finalize("multifamily_new_construction", excluded, reasoning, pilotArchetypes);
  }

  // Mixed-use new construction
  if (scope.mixed_occupancy && !tiCues.test(planText)) {
    reasoning.push("Mixed-occupancy declaration without TI markers — likely new construction");
    return finalize("mixed_use_new_construction", excluded, reasoning, pilotArchetypes);
  }

  // Fallback: we couldn't confidently bucket it
  reasoning.push("Could not classify into a known archetype — manual triage required");
  return finalize("unclassified", excluded, reasoning, pilotArchetypes);
}

function finalize(
  archetype: ProjectArchetype,
  excluded: string[],
  reasoning: string[],
  pilotArchetypes?: ProjectArchetype[],
): ArchetypeResult {
  // If the agency hasn't set a pilot allowlist, accept everything that
  // would otherwise be IN_SCOPE. If it has set one, the allowlist wins.
  const allow = pilotArchetypes && pilotArchetypes.length > 0
    ? new Set(pilotArchetypes)
    : IN_SCOPE;
  const inScope = allow.has(archetype);
  return {
    archetype,
    in_pilot_scope: inScope,
    reasoning,
    excluded_overlays: excluded,
  };
}

// =====================================================================
// Human-readable reason for the dashboard / comment letter
// =====================================================================
export function renderArchetypeBanner(result: ArchetypeResult): string {
  if (result.in_pilot_scope) {
    return `In-pilot archetype: ${result.archetype}. AI triage proceeded.`;
  }
  const why = result.excluded_overlays.length
    ? result.excluded_overlays.join("; ")
    : result.reasoning[result.reasoning.length - 1] ?? "Out of current pilot scope";
  return `OUT OF PILOT SCOPE (${result.archetype}). Reason: ${why}. Send to manual review.`;
}

File: backend/app/agents/archetype.py
"""Project archetype classifier — intake out-of-scope gate.

Ported from plan-room-ahj/supabase/functions/_shared/archetype.ts. Runs
at intake to decide whether a submittal is inside the current pilot
scope. Out-of-scope plans get rejected before the rule engine even
runs — keeping the accuracy claim defensible.

"Pilot scope" is an archetype allowlist (see config/pilot.py). Pure
function — no LLM, no network. Inputs are the extracted plan data,
the resolved property overlay (if any), and the raw plan text for
keyword cues. Reasoning is human-readable so reviewers can see why
something was excluded.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import List, Optional, Sequence

from app.config.pilot import (
    ARCHETYPE_HIGH_RISE_OR_MID_RISE,
    ARCHETYPE_LA_COASTAL_ZONE,
    ARCHETYPE_LA_HILLSIDE_SFR,
    ARCHETYPE_LA_HPOZ_PROPERTY,
    ARCHETYPE_LA_SFR_TYP_VB_MINISTERIAL,
    ARCHETYPE_LA_TI_COMMERCIAL,
    ARCHETYPE_MIXED_USE_NEW_CONSTRUCTION,
    ARCHETYPE_MULTIFAMILY_NEW_CONSTRUCTION,
    ARCHETYPE_UNCLASSIFIED,
    ARCHETYPE_VENTURA_AG_BUILDING,
    ARCHETYPE_VENTURA_SFR_TYP_VB_MINISTERIAL,
    ARCHETYPE_VENTURA_TI_COMMERCIAL,
    ARCHETYPE_VENTURA_VHFHSZ_SFR,
    PILOT_ARCHETYPES_DEFAULT,
)
from app.models.schemas import ExtractedPlanData


# =====================================================================
# Property profile — lightweight, optional. Populated from GIS overlay
# lookups (services/gis_overlays.py via site_resolver) when the customer
# provided a project address at upload; absent otherwise, in which case
# the gate falls back to plan-text cues, which is still valuable.
# =====================================================================
@dataclass
class ParcelInfo:
    jurisdiction: Optional[str] = None
    zoning_code: Optional[str] = None


@dataclass
class CoastalZoneInfo:
    in_coastal_zone: bool = False


@dataclass
class WuiInfo:
    in_wui: bool = False
    haz_class: Optional[str] = None  # e.g. "very_high"


@dataclass
class PropertyProfile:
    parcel: Optional[ParcelInfo] = None
    coastal_zone: Optional[CoastalZoneInfo] = None
    wui_zone: Optional[WuiInfo] = None
    # GIS-resolved LA-city overlay flags. None = not checked (no address /
    # layer failed) — only an explicit True triggers a reject, so unknown
    # never masquerades as "in zone".
    in_hpoz: Optional[bool] = None
    in_hillside: Optional[bool] = None


# =====================================================================
# Result
# =====================================================================
@dataclass
class ArchetypeResult:
    archetype: str
    in_pilot_scope: bool
    reasoning: List[str] = field(default_factory=list)
    excluded_overlays: List[str] = field(default_factory=list)


# =====================================================================
# Plan-text regex cues
# =====================================================================
_VENTURA_TEXT_CUE = re.compile(
    r"\bventura\s+county\b|\bcounty\s+of\s+ventura\b|\bcamarillo\b|"
    r"\bsimi\s+valley\b|\bthousand\s+oaks\b|\boxnard\b|\bojai\b|"
    r"\bmoorpark\b|\bport\s+hueneme\b|\bfillmore\b|\bsanta\s+paula\b",
    re.IGNORECASE,
)
_WUI_TEXT_CUE = re.compile(
    r"\bvery\s+high\s+fire\s+hazard\s+severity\s+zone\b|\bVHFHSZ\b|"
    r"\bWUI\b|\bChapter\s+7A\b",
    re.IGNORECASE,
)
_AG_TEXT_CUE = re.compile(
    r"\bagricultural\s+(building|structure|exempt)|\bA-E\b|\bAE-\d|"
    r"\bopen\s+space\s+agricultural\b",
    re.IGNORECASE,
)
_HILLSIDE_CUES = re.compile(
    r"\bbaseline\s+hillside|\bBHO\b|\bhillside\s+ordinance\b|"
    r"\bmulholland\b.*\bspecific\s+plan|RE40-?1H|RE15-?1H",
    re.IGNORECASE,
)
_HPOZ_CUES = re.compile(
    r"\bHPOZ\b|\bhistoric\s+preservation\s+overlay",
    re.IGNORECASE,
)
_TI_CUES = re.compile(
    r"\btenant\s+improvement\b|\bTI\b\s+plans?|interior\s+alteration|"
    r"interior\s+remodel|\bsuite\s+\d",
    re.IGNORECASE,
)
_HILLSIDE_ZONING_CODE = re.compile(r"H$|HILLSIDE|\bBMO\b|\bBHO\b", re.IGNORECASE)
# A plan that declares a detached single-family dwelling. R-3 occupancy and
# Type V-B construction are the default for these, so this cue lets us classify
# an obvious SFR even when the title sheet never spells out "R-3"/"V-B" in
# selectable text (common on residential sets).
_SFR_TEXT_CUE = re.compile(
    r"\bsingle[\s-]*family\s+(dwelling|residence|home|house)\b|"
    r"\bone[\s-]*family\s+dwelling\b|\bSFD\b|\bSFR\b|"
    r"\b(new|proposed|\(n\))\s+single[\s-]*family\b",
    re.IGNORECASE,
)


# =====================================================================
# classify_archetype
#
# Order matters: we check "kicked out by overlay" cases first (cheapest
# reject), then identify the in-scope archetype, then fall back to
# unclassified.
# =====================================================================
def classify_archetype(
    scope: ExtractedPlanData,
    plan_text: str,
    property_profile: Optional[PropertyProfile] = None,
    pilot_archetypes: Optional[Sequence[str]] = None,
) -> ArchetypeResult:
    """Decide whether this submittal is in the pilot's allowlist.

    Args:
        scope: The Surveyor's extracted plan data.
        plan_text: Raw extracted text from the plan PDF (for keyword cues).
        property_profile: Optional GIS-resolved property overlays. When
            absent, the gate falls back to plan-text cues + scope.wui_zone.
        pilot_archetypes: Per-agency allowlist override. Falls back to
            PILOT_ARCHETYPES_DEFAULT when None or empty.

    Returns:
        ArchetypeResult with the assigned archetype, in_pilot_scope
        verdict, human-readable reasoning, and any overlay flags that
        kicked the case out.
    """
    reasoning: List[str] = []
    excluded: List[str] = []

    # ---- Jurisdiction detection (plan-text + parcel hint) --------------
    parcel = property_profile.parcel if property_profile else None
    is_ventura = bool(
        (parcel and parcel.jurisdiction and re.search(r"ventura", parcel.jurisdiction, re.IGNORECASE))
        or _VENTURA_TEXT_CUE.search(plan_text or "")
    )

    # ---- Hard rejects: parcel-derived overlays -------------------------
    if parcel and parcel.zoning_code:
        z = parcel.zoning_code.upper()
        if _HILLSIDE_ZONING_CODE.search(z):
            excluded.append("Hillside / BHO zoning")
            reasoning.append(f'Parcel zoning code "{z}" indicates Hillside overlay')
            return _finalize(ARCHETYPE_LA_HILLSIDE_SFR, excluded, reasoning, pilot_archetypes)

    if property_profile and property_profile.in_hillside:
        excluded.append("Hillside Ordinance (GIS)")
        reasoning.append("Address falls inside the LA Hillside Ordinance area (GIS overlay)")
        return _finalize(ARCHETYPE_LA_HILLSIDE_SFR, excluded, reasoning, pilot_archetypes)

    if property_profile and property_profile.in_hpoz:
        excluded.append("HPOZ (GIS)")
        reasoning.append("Address falls inside a Historic Preservation Overlay Zone (GIS overlay)")
        return _finalize(ARCHETYPE_LA_HPOZ_PROPERTY, excluded, reasoning, pilot_archetypes)

    if property_profile and property_profile.coastal_zone and property_profile.coastal_zone.in_coastal_zone:
        # Coastal is an in-pilot archetype (allowlist) since the Coastal Act +
        # LCP corpus layers landed — classify it, but don't mark it excluded.
        # Agencies that override the allowlist without it still get a clean
        # out-of-scope verdict from _finalize.
        reasoning.append(
            "Address is inside the CA Coastal Zone — coastal code stack "
            "(Coastal Act + certified LCP) applies"
        )
        return _finalize(ARCHETYPE_LA_COASTAL_ZONE, excluded, reasoning, pilot_archetypes)

    # ---- Ventura County VHFHSZ reject ----------------------------------
    wui = property_profile.wui_zone if property_profile else None
    if is_ventura and wui and wui.in_wui:
        haz = wui.haz_class or "very_high"
        excluded.append(f"Ventura VHFHSZ ({haz})")
        reasoning.append(
            f"Address falls in CalFire {haz} FHSZ — CBC Ch. 7A wildfire-resistive "
            "materials review out of pilot scope"
        )
        return _finalize(ARCHETYPE_VENTURA_VHFHSZ_SFR, excluded, reasoning, pilot_archetypes)

    # Fall back to scope.wui_zone (set by surveyor when address is known)
    # plus plan-text cue, since ai-plan-checker doesn't yet populate
    # PropertyProfile.
    scope_wui = getattr(scope, "wui_zone", None)
    if is_ventura and scope_wui in ("very_high", "high"):
        excluded.append(f"Ventura VHFHSZ ({scope_wui})")
        reasoning.append(
            f"Address is flagged as {scope_wui} FHSZ — CBC Ch. 7A out of pilot scope"
        )
        return _finalize(ARCHETYPE_VENTURA_VHFHSZ_SFR, excluded, reasoning, pilot_archetypes)

    if is_ventura and _WUI_TEXT_CUE.search(plan_text or ""):
        excluded.append("Ventura VHFHSZ (plan-text cue)")
        reasoning.append("Plan text references Ventura County WUI / Very-High FHSZ")
        return _finalize(ARCHETYPE_VENTURA_VHFHSZ_SFR, excluded, reasoning, pilot_archetypes)

    # ---- Ventura County agricultural building reject -------------------
    if is_ventura and _AG_TEXT_CUE.search(plan_text or ""):
        excluded.append("Ventura agricultural building")
        reasoning.append(
            "Plan text indicates an agricultural-zoned structure "
            "(county-specific exemptions apply)"
        )
        return _finalize(ARCHETYPE_VENTURA_AG_BUILDING, excluded, reasoning, pilot_archetypes)

    # ---- Plan-text overlay cues (when no property profile) -------------
    if _HILLSIDE_CUES.search(plan_text or ""):
        excluded.append("Hillside / BHO (plan-text cue)")
        reasoning.append("Plan text references Baseline Hillside Ordinance or Hillside zoning")
        return _finalize(ARCHETYPE_LA_HILLSIDE_SFR, excluded, reasoning, pilot_archetypes)

    if _HPOZ_CUES.search(plan_text or ""):
        excluded.append("HPOZ")
        reasoning.append("Plan text references HPOZ (Historic Preservation Overlay Zone)")
        return _finalize(ARCHETYPE_LA_HPOZ_PROPERTY, excluded, reasoning, pilot_archetypes)

    # ---- Building scale rejects ----------------------------------------
    height = getattr(scope, "building_height", None)
    if height is not None and height > 75:
        excluded.append("High-rise (> 75 ft)")
        reasoning.append(f"Building height {height} ft exceeds 75 ft high-rise threshold")
        return _finalize(ARCHETYPE_HIGH_RISE_OR_MID_RISE, excluded, reasoning, pilot_archetypes)

    stories = getattr(scope, "stories", None)
    if stories is not None and stories >= 5:
        excluded.append("Mid-rise (>= 5 stories)")
        reasoning.append(
            f"{stories} stories puts this above the pilot single-family / TI scope"
        )
        return _finalize(ARCHETYPE_HIGH_RISE_OR_MID_RISE, excluded, reasoning, pilot_archetypes)

    # ---- In-scope identification ---------------------------------------
    has_ti_cue = bool(_TI_CUES.search(plan_text or ""))

    occupancy = (getattr(scope, "occupancy_type", None) or "").upper().strip()
    construction = (getattr(scope, "construction_type", None) or "").upper().strip()

    # Detect mixed occupancy via "+" or "/" in the occupancy_type string
    # (ai-plan-checker doesn't have a dedicated mixed_occupancy flag yet).
    mixed_occupancy = bool(occupancy and re.search(r"[+/]| AND |,", occupancy))

    # TI = explicit "tenant improvement" / interior alteration cues
    if has_ti_cue and not mixed_occupancy:
        if is_ventura:
            reasoning.append("Ventura County commercial tenant improvement")
            return _finalize(
                ARCHETYPE_VENTURA_TI_COMMERCIAL, excluded, reasoning, pilot_archetypes
            )
        reasoning.append("Plan text identifies this as a commercial tenant improvement")
        return _finalize(ARCHETYPE_LA_TI_COMMERCIAL, excluded, reasoning, pilot_archetypes)

    # SFR Type V-B = R-3, single occupancy, V-B construction, modest scale
    is_r3 = "R-3" in occupancy or occupancy == "R-3"
    is_vb = construction in ("V-B", "VB", "TYPE V-B", "V B")

    per_story = getattr(scope, "per_story_area", None)
    building_area = getattr(scope, "building_area", None)
    area_used = per_story if per_story is not None else (building_area or 0)
    modest_area = area_used <= 10_000

    low_rise = stories is None or stories <= 3

    if is_r3 and is_vb and modest_area and low_rise:
        if is_ventura:
            reasoning.append(
                "Ventura County single-family R-3 / Type V-B, <= 3 stories, "
                "no VHFHSZ or ag overlay"
            )
            return _finalize(
                ARCHETYPE_VENTURA_SFR_TYP_VB_MINISTERIAL,
                excluded,
                reasoning,
                pilot_archetypes,
            )
        reasoning.append("Single-family R-3 / Type V-B with modest area and <= 3 stories")
        return _finalize(
            ARCHETYPE_LA_SFR_TYP_VB_MINISTERIAL, excluded, reasoning, pilot_archetypes
        )

    # Robust SFR fallback: a detached single-family dwelling is R-3 / Type V-B
    # by default. Many residential title sheets never put the literal "R-3" /
    # "V-B" strings in selectable text (and vision can't read a code-data box
    # that isn't on the sheet), so requiring them rejects obviously in-scope
    # homes. When the plan text declares a single-family dwelling, there's no
    # TI / multifamily / mixed-occupancy signal, and it's low-rise + modest
    # area, treat it as the in-scope SFR archetype. The overlay rejects
    # (hillside, HPOZ, coastal, fire-zone, high/mid-rise) already ran above, so
    # we can't mislabel one of those as a plain SFR here.
    sfr_by_text = bool(_SFR_TEXT_CUE.search(plan_text or ""))
    bigger_use = mixed_occupancy or "R-1" in occupancy or "R-2" in occupancy
    if sfr_by_text and not has_ti_cue and not bigger_use and modest_area and low_rise:
        note = (
            " (occupancy/construction not explicit in the plan text — inferred "
            "the R-3 / Type V-B default for a detached dwelling)"
        )
        if is_ventura:
            reasoning.append("Ventura County single-family dwelling declared in plan text" + note)
            return _finalize(
                ARCHETYPE_VENTURA_SFR_TYP_VB_MINISTERIAL, excluded, reasoning, pilot_archetypes
            )
        reasoning.append("Single-family dwelling declared in plan text" + note)
        return _finalize(
            ARCHETYPE_LA_SFR_TYP_VB_MINISTERIAL, excluded, reasoning, pilot_archetypes
        )

    # Multifamily new construction = R-1 or R-2 + no TI cue
    if ("R-1" in occupancy or "R-2" in occupancy) and not has_ti_cue:
        reasoning.append("R-1 / R-2 occupancy without TI markers — multifamily new construction")
        return _finalize(
            ARCHETYPE_MULTIFAMILY_NEW_CONSTRUCTION, excluded, reasoning, pilot_archetypes
        )

    # Mixed-use new construction
    if mixed_occupancy and not has_ti_cue:
        reasoning.append("Mixed-occupancy declaration without TI markers — likely new construction")
        return _finalize(
            ARCHETYPE_MIXED_USE_NEW_CONSTRUCTION, excluded, reasoning, pilot_archetypes
        )

    # Fallback: we couldn't confidently bucket it
    reasoning.append("Could not classify into a known archetype — manual triage required")
    return _finalize(ARCHETYPE_UNCLASSIFIED, excluded, reasoning, pilot_archetypes)


# =====================================================================
# Internal helpers
# =====================================================================
def _finalize(
    archetype: str,
    excluded: List[str],
    reasoning: List[str],
    pilot_archetypes: Optional[Sequence[str]],
) -> ArchetypeResult:
    """Bind an archetype label to its in_pilot_scope verdict.

    If the agency hasn't set a pilot allowlist, accept everything that
    would otherwise be in PILOT_ARCHETYPES_DEFAULT. If it has set one,
    the agency allowlist wins.
    """
    allow = pilot_archetypes if pilot_archetypes else PILOT_ARCHETYPES_DEFAULT
    return ArchetypeResult(
        archetype=archetype,
        in_pilot_scope=archetype in allow,
        reasoning=reasoning,
        excluded_overlays=excluded,
    )


# =====================================================================
# Human-readable reason for the dashboard / comment letter
# =====================================================================
def render_archetype_banner(result: ArchetypeResult) -> str:
    """One-line summary for the reviewer dashboard."""
    if result.in_pilot_scope:
        return f"In-pilot archetype: {result.archetype}. AI triage proceeded."
    why = (
        "; ".join(result.excluded_overlays)
        if result.excluded_overlays
        else (result.reasoning[-1] if result.reasoning else "Out of current pilot scope")
    )
    return (
        f"OUT OF PILOT SCOPE ({result.archetype}). Reason: {why}. Send to manual review."
    )
